### ISSUE TITLE
Question about how the interface suggest variables/constants

### ISSUE THREAD

**Issue by yurimalheiros** [julianmendez/uel#6]
Hi, this is a question, not a particular issue with the software.

I would like to know, given a equation and a background ontology, how the standalone interface chooses which variables/constants to suggest to the user. Besides that, is there any problem, using the library in the Java code, to pass some variable that the interface would not suggest?

Thank you.

**Comment by stefborg** [comment]
Hi Yuri,

First, every OWL class that appears in the goal (subsumptions, equations, dissubsumptions, disequations), but has no definition in the background ontology, can be marked as a variable.

Classes that are defined in the background ontology, either with a primitive definition (subclass axiom) or a full definition (equivalent classes axiom), are marked as "internal" variables, since the definitions are automatically imported into the unification problem. For primitive definitions, we introduce additional "UNDEF" classes to make them full definitions. These UNDEF classes can also be marked as (user) variables, since we do not know a priori whether they should be constants or variables.

In the user interface, there is currently no way to mark other classes as variables. Using the library, however, you can pass any OWL class, and hence for example also make one of the defined classes a user variable. The difference to an internal variable is that the substitution of user variables will be shown in the unifier. However, usually the defined classes are not interesting to see in the unifier, since their definition in the background ontology places a tight restriction on their substitution.

Did this answer your question?

Bye, Stefan

**Comment by yurimalheiros** [comment]
Hi Stephan, thank you for your quick answer.

I am not 100% sure if I understood everything, so, I would like to make other questions.

Given a background ontology with these axioms:

X ≡ A ⋂ ∃ r.B
Y ≡ A ⋂ ∃ r.C

(A, B, and C have no definition)

If I define an equation X ≡ Y, to unify, the interface will suggest A, B, and C, as constants/variables.
In this case, X, and Y, the concepts that appear in the goal have definitions, thus they can't be suggested by the interface.

So, does the interface also consider the concepts in the definitions of the concepts in the equation?
For example, X is a concept in the equation, but the interface suggest A and B as constants/variables.

What happened if, in the example above, I added the axiom B ≡ D in the background ontology? The interface stop to suggest B, right?

I do not know if I was clear enough in these new questions, then if you need further explanation about my questions, please tell me.

Thanks again, Stephan.

**Comment by stefborg** [comment]
Hi Yuri,

Yes, UEL expands all definitions from the background ontology recursively (only those that are needed) and adds them as equations to the unification problem. Afterwards, if you have an equation like X ≡ A ⋂ ∃ r.B, then it does not make sense to make X a constant, since then this equation could not be solved anyway. Thus, we mark it automatically as a variable, and the user cannot choose it anymore.

But even if it was a user-defined variable, its substitution would not be interesting since it is forced to be equivalent to A ⋂ ∃ r.B, and hence can be determined from the substitutions of A and B (if they are variables). Thus, we do not show the substitutions of such automatically chosen internal variables to the user. As I said, you should be able to override this behavior when using the library and adding X (with full IRI!) to the set of variables.

In your example above, if you add B ≡ D to the background ontology, you are correct that the user cannot choose B anymore as variable/constant, but instead D.

Bye, Stefan

**Comment by yurimalheiros** [comment]
Hi Stephan, thank you so much for taking your time to explain this to me.
I think now I understood how the selection of concepts to be constants/variables works.

**Comment by yurimalheiros** [comment]
Hello Stephan.

I was testing the library and I got an unexpected behavior setting a variable with a definition (subclass definition or equivalence definition).

First, I defined this ontology:

<img width="183" alt="captura de tela 2016-04-24 14 05 55" src="https://cloud.githubusercontent.com/assets/161663/14769126/12571472-0a28-11e6-8aa0-5bfed4b58545.png">

In this picture you can see the subclass/superclass relations of the concepts. Also, I defined the following axioms:

FastCar ≡ Car ⋂ ∃ hasEngine.FastEngine
RaceCar ≡ Car ⋂ ∃ hasEngine.RaceCar

Note that, the axioms FastCar ⊆ Car and RaceCar ⊆ Car do not exist, the hierarchy among these concepts is defined only by the two equivalence axioms above.

So, I used this ontology as the background ontology, set the concept Car as the variable, and the subsumption RaceCar ⊆ FastCar (equations, disequations, and dissubsumptions are empty). After I ran the library (version 1.3.1) with SAT_BASED_ALGORITHM_MINIMAL. Then I got this result:

```
EquivalentClasses(<http://www.semanticweb.org/yurimalheiros/ontologies/2016/3/untitled-ontology-121#FastEngine_UNDEF> owl:Thing )

EquivalentClasses(<http://www.semanticweb.org/yurimalheiros/ontologies/2016/3/untitled-ontology-121#Car_UNDEF> owl:Thing )

EquivalentClasses(<http://www.semanticweb.org/yurimalheiros/ontologies/2016/3/untitled-ontology-121#RaceEngine_UNDEF> owl:Thing )
```

I noticed that the variable Car was ignored, even I defined it as my variable. I think, maybe because Car has a definition in the background ontology. After that, I removed the definition of Car and tested again.

The following image show the new ontology:

<img width="173" alt="captura de tela 2016-04-24 14 06 01" src="https://cloud.githubusercontent.com/assets/161663/14769067/a67566ce-0a26-11e6-81ae-c72a5e5a04a1.png">

(FastCar and RaceCar have the same definitions)

I ran the library again, and I got this result:

```
EquivalentClasses(<http://www.semanticweb.org/yurimalheiros/ontologies/2016/3/untitled-ontology-121#FastEngine_UNDEF> owl:Thing )

EquivalentClasses(<http://www.semanticweb.org/yurimalheiros/ontologies/2016/3/untitled-ontology-121#Car> owl:Thing )

EquivalentClasses(<http://www.semanticweb.org/yurimalheiros/ontologies/2016/3/untitled-ontology-121#RaceEngine_UNDEF> owl:Thing )
```

Now the concept Car appears in the results (second line).

After that, I tested again with more changes in the background ontology. I removed the subclass definition of FastEngine and RaceEngine. See image below:

<img width="158" alt="captura de tela 2016-04-24 14 06 15" src="https://cloud.githubusercontent.com/assets/161663/14769152/d4b7e49c-0a28-11e6-8669-e366a31dc39e.png">

I ran the library again, and now I got a interesting unifier.

```
EquivalentClasses(<http://www.semanticweb.org/yurimalheiros/ontologies/2016/3/untitled-ontology-121#Car> ObjectSomeValuesFrom(<http://www.semanticweb.org/yurimalheiros/ontologies/2016/3/untitled-ontology-121#hasEngine> <http://www.semanticweb.org/yurimalheiros/ontologies/2016/3/untitled-ontology-121#FastEngine>) )
```

Now I am very curious about why this unifier does not appeared in my second test. Because, it solves the subsumption RaceCar ⊆ FastCar too. 

Summing up, I have two questions:

- Is it not possible to use a variable that has some kind of definition (equivalence or subclass) even in the library?
- Do you know why I did not get the unifier presented in the third test in my second test?

Sorry for the long message and all these questions, but I am really interested in use your library in my research.

Thanks again, Stephan.

**Comment by stefborg** [comment]
Hi Yuri,

OK, I was wrong. You cannot mark defined classes as user variables in 1.3.1. I do not recall this specific change, but in the new version 1.4.0 this is possible. I recommend that you try this version (but I cannot guarantee that we did not break anything else).

Nevertheless, even if Car is not a user variable, then you know that it is equivalent to Car_UNDEF (or "Car_UNDEF ⊓ Vehicle" in the first ontology). Hence, from the substitution of Car_UNDEF you can always reconstruct the one of Car.

The reason that you did not get the last unifier also in the second run is probably that you use SAT_BASED_ALGORITHM_MINIMAL. The substitution where everything is mapped to owl:Thing is a unifier, and it is certainly minimal, so the algorithm does not bother to look for other solutions. There are two options here:

1. You can use SAT_BASED_ALGORITHM. This will probably give you also the minimal one first, but you should also get the desired one later. The downside is that you may get thousands of unifiers, if not more, and may have to search through all of them.

2. To solve the above problem, we introduced disunification. The first thing I would try is to introduce a dissubsumption "owl:Thing ⋢ Car", which forces at least one non-variable atom to be in the substitution of Car (with Car_UNDEF instead, you should get similar results). The downside here is that it may take a little longer to solve, and that the unifiers are not always what one would first expect (at least this was the case in our experiments). I would be very interested in the results of both SAT_BASED_ALGORITHM and SAT_BASED_ALGORITHM_MINIMAL in this case.

Bye, Stefan

**Comment by yurimalheiros** [comment]
I see. Very good explanation again. The "owl:Thing ⋢ Car" really helped me. I will continue to work with the library, and if I have more questions I will let you know.

Thank you very much!
